### ISSUE TITLE
Error while rendering template in Python 3.11

### ISSUE THREAD

**Issue by vladimir-poghosyan** [akornatskyy/wheezy.template#68]
After trying to upgrade Python from 3.10.8 to 3.11.0 I have encountered the following error during template rendering:

```
Traceback (most recent call last):                                                                                                                          
  File "/usr/local/lib/python3.11/site-packages/wheezy/template/engine.py", line 70, in render                                                               
    return self.renders[name](ctx, local_defs, super_defs)                                                                                                   
           ~~~~~~~~~~~~^^^^^^                                                                                                                                
KeyError: 'errors/http_500.html'                                                                                                                            
During handling of the above exception, another exception occurred:                                                                                         
Traceback (most recent call last):                                                                                                                          
  File "/usr/local/lib/python3.11/site-packages/wheezy/web/middleware/errors.py", line 25, in __call__                                                       
    response = following(request)                                                                                                                            
               ^^^^^^^^^^^^^^^^^^                                                                                                                            
  File "/usr/local/lib/python3.11/site-packages/wheezy/http/application.py", line 20, in <lambda>                                                            
    return lambda request: func(request, following)                                                                                                          
                           ^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                          
  File "/usr/local/lib/python3.11/site-packages/wheezy/web/middleware/routing.py", line 24, in __call__                                                      
    return handler(request)                                                                                                                                  
           ^^^^^^^^^^^^^^^^                                                                                                                                  
  File "/usr/local/lib/python3.11/site-packages/wheezy/web/handlers/template.py", line 24, in handle_request                                                 
    return h()                                                                                                                                               
           ^^^                                                                                                                                               
  File "/usr/local/lib/python3.11/site-packages/wheezy/web/handlers/method.py", line 29, in __call__                                                         
    response = self.get()                                                                                                                                    
               ^^^^^^^^^^                                                                                                                                    
  File "/usr/local/lib/python3.11/site-packages/wheezy/web/handlers/template.py", line 33, in get                                                            
    response = self.render_response(self.template_name)                                                                                                      
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                      
  File "/usr/local/lib/python3.11/site-packages/wheezy/web/handlers/base.py", line 129, in render_response                                                   
    options["render_template"](                                                                                                                              
  File "/usr/local/lib/python3.11/site-packages/wheezy/web/templates.py", line 137, in __call__                                                              
    return self.render(template_name, kwargs, {}, {})                                                                                                        
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                        
  File "/usr/local/lib/python3.11/site-packages/wheezy/template/engine.py", line 72, in render                                                               
    self.compile_template(name)                                                                                                                              
  File "/usr/local/lib/python3.11/site-packages/wheezy/template/engine.py", line 110, in compile_template
    render_template = self.compiler.compile_source(           
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^           
  File "/usr/local/lib/python3.11/site-packages/wheezy/template/compiler.py", line 27, in compile_source
    compiled = compile(node, name, "exec")                    
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^                    
ValueError: AST node column range (0, 55) for line range (-1, 22) is not valid
```
I am using official docker image for Python (python:3.11.0-slim-bullseye).

Thanks.

**Comment by akornatskyy** [comment]
the issue is potentially related to [ast.increment_lineno](https://docs.python.org/3.11/library/ast.html#ast.increment_lineno). it used to adjust source line number by -2 (so template line number match lines in python code).

Potentially, as a workaround you can try the following:

```py
engine.compiler.source_lineno = 0;
```

**Comment by vladimir-poghosyan** [comment]
Thank you for quick response.

I have tried as you suggested and everything worked just fine. Should this be considered as a permanent solution?

**Comment by akornatskyy** [comment]
I am not certain that this issue with [ast.increment_lineno](https://docs.python.org/3.11/library/ast.html#ast.increment_lineno) is supposed to intentionally raise error (because a good reason behind, which I don't know) or that it's not and will return back to previous behavior (in one of patches to the standard library later)... which likely depends on a number of "complains" from the community.

**Comment by char101** [comment]
The ast source is offseted by -2 because the compiler adds two lines before the template code

```python
def render(ctx, local_defs, super_defs):
    _b = []; w = _b.append
```

but this causes the ast lineno become negative (1-2 = -1), and in 3.11 this will thrown an error (https://github.com/python/cpython/issues/93351)

https://github.com/python/cpython/blob/2b38a9aa747785f6d540e95e6846d6510de6b306/Python/ast.c#L32

I think rather than offseting the ast lineno, it is better to offset the lineno when reporting errors.

**Comment by akornatskyy** [comment]
right, now try to evaluate the effort to get it done the better way.

**Comment by char101** [comment]
If we set

```
engine.compiler.source_lineno = 0
```

on render error, the displayed line in the template will be wrong. To fix the lineno in the exception we can [rewrite the traceback](https://github.com/pallets/jinja/blob/main/src/jinja2/debug.py#L14) like this

```python
import sys
from jinja2.debug import fake_traceback

def render_wheezy(template, vars):
    try:
        return template.render(vars)
    except Exception as ex:
        stack = []
        tb = sys.exc_info()[2]
        while tb is not None:
            file = tb.tb_frame.f_code.co_filename
            if file and not file.endswith('.py'):
                stack.append(fake_traceback(ex, tb, file, tb.tb_lineno - 2))
            else:
                stack.append(tb)
            tb = tb.tb_next

        tb_next = None
        for tb in reversed(stack):
            tb.tb_next = tb_next
            tb_next = tb
        raise ex.with_traceback(tb_next)
```

**Comment by MathieuDuponchelle** [comment]
@akornatskyy could we get a fix for this ideally? :)

**Comment by akornatskyy** [comment]
right now there is a reasonable workaround:

```python
engine.compiler.source_lineno = 0
```

I am not certain if that is kind of feature now in python 3.11 or they will revert it back to previous behavior; so keeping it open for some time.

**Comment by char101** [comment]
If you read the python commit I linked above, it is part of the new precise error reporting in python 3.11 and will surely not be reverted. Having an ast with negative lineno in the first place is wrong.

**Comment by MathieuDuponchelle** [comment]
> right now there is a reasonable workaround:

Oh I see, I'll switch to that then thanks, but if @char101 is right and you have another solution in mind already, maybe you can implement it? No pressure of course, thanks for being reactive here :)

**Comment by AndydeCleyre** [comment]
> right now there is a reasonable workaround:

Is there some way to apply this when using `wheezy.template` as a CLI tool, rather than a library?

**Comment by char101** [comment]
> Is there some way to apply this when using `wheezy.template` as a CLI tool, rather than a library?

You can put it in `usercustomize.py`

```python
> import site
> print(site.USER_SITE)
```

Create `usercustomize.py` in that directory and put above line in that file.

**Comment by akornatskyy** [comment]
a [workaround](https://github.com/akornatskyy/wheezy.template/commit/639128b7107961c1a11b2615886baac7507df47a) applied and released with version `3.2.0`.
